xds/internal/resolver/xds_resolver.go

```diff
@@ -236,94 +284,248 @@ func (r *xdsResolver) sendNewServiceConfig(cs *configSelector) bool {
+}
+
+// onResolutionComplete performs the following actions when the resolver has
+// received sufficient configuration from the xDS client:
+//   - creates a new config selector
+//   - prunes active clusters and pushes a new service config on the channel
+//   - decrement references to the old config selector and assigns the new one
+//     as the current one.
+//
+// Only executed in the context of a serializer callback.
+func (r *xdsResolver) onResolutionComplete() {
+	if !r.resolutionComplete() {
 		return
 	}
-	// Remove any existing entry in updateCh and replace with the new one.
-	select {
-	case <-r.updateCh:
-	default:
+
+	cs, err := r.newConfigSelector()
+	if err != nil {
+		r.logger.Warningf("Failed to build a config selector for resource %q: %v", r.ldsResourceName, err)
+		r.cc.ReportError(err)
+		return
+	}
+
+	if !r.sendNewServiceConfig(cs) {
+		// JSON error creating the service config (unexpected); erase
+		// this config selector and ignore this update, continuing with
+		// the previous config selector.
+		cs.stop()
+		return
 	}
-	r.updateCh <- suWithError{su: su, err: err}
+
+	r.curConfigSelector.stop()
+	r.curConfigSelector = cs
 }
 
-// ResolveNow is a no-op at this point.
-func (*xdsResolver) ResolveNow(o resolver.ResolveNowOptions) {}
+func (r *xdsResolver) applyRouteConfigUpdate(update xdsresource.RouteConfigUpdate) {
+	matchVh := xdsresource.FindBestMatchingVirtualHost(r.ldsResourceName, update.VirtualHosts)
+	if matchVh == nil {
+		// No matching virtual host found.
+		r.onError(fmt.Errorf("no matching virtual host found for %q", r.ldsResourceName))
+		return
+	}
+	r.currentRouteConfig = update
+	r.currentVirtualHost = matchVh
+	r.routeConfigUpdateRecvd = true
 
-// Close closes the resolver, and also closes the underlying xdsClient.
-func (r *xdsResolver) Close() {
-	// Note that Close needs to check for nils even if some of them are always
-	// set in the constructor. This is because the constructor defers Close() in
-	// error cases, and the fields might not be set when the error happens.
-	if r.cancelWatch != nil {
-		r.cancelWatch()
+	r.onResolutionComplete()
+}
+
+func (r *xdsResolver) onListenerResourceUpdate(update xdsresource.ListenerUpdate) {
+	r.currentListener = update
+	r.listenerUpdateRecvd = true
+
+	if update.InlineRouteConfig != nil {
+		// If there was a previous route config watcher because of a non-inline
+		// route configuration, cancel it.
+		r.rdsResourceName = ""
+		if r.routeConfigWatcher != nil {
+			r.routeConfigWatcher.stop()
+			r.routeConfigWatcher = nil
+		}
+
+		r.applyRouteConfigUpdate(*update.InlineRouteConfig)
+		return
 	}
-	if r.xdsClientClose != nil {
-		r.xdsClientClose()
+
+	// We get here only if there was no inline route configuration.
+
+	// If the route config name has not changed, send an update with existing
+	// route configuration and the newly received listener configuration.
+	if r.rdsResourceName == update.RouteConfigName {
+		r.onResolutionComplete()
```

Comment:
If we really don't want to nil matchVh here, can you write a comment about using old configuration in the error case (similar to your comments on onError()).